docs/design/2024-12-20-ticdc-flow-control.md

```diff
@@ -0,0 +1,72 @@
+# TiCDC Flow Control Design
+
+- Author(s): [dongmen](https://github.com/asddongmen)
+- Tracking Issue(s): [736](https://github.com/pingcap/ticdc/issues/736)
+
+## Background
+
+TiCDC processes data in two main parts:
+
+1. **Data Puller**: Includes LogPuller, EventStore, and EventService, responsible for fetching upstream data.
+2. **Data Sinker**: Includes EventCollector, Dispatcher, and Sink, responsible for writing data to downstream systems.
+
+The following diagram illustrates the relationship between the **data puller** and **data sinker**:
+
+![Data Flow](./medias/flow-control-1.png)
+<!-- The source file for this diagram: docs/design/medias/flow-control-1.puml -->
+
+In this architecture, **EventService** and **EventCollector** facilitate communication between the two parts:  
+
+- **EventCollector** initiates a subscription and establishes a data stream.  
+
+- The data stream is buffered in **EventCollector**'s memory before being processed by other components.
+
+When downstream consumption lags behind upstream production, data accumulates in memory, potentially causing OOM. To address this, a flow control mechanism is necessary to stop pushing data when congestion occurs.
+
+---
+
+## Design Details
+
+### Pause and Resume Mechanism
+
+The **EventCollector** uses the `dynamicStream` component to handle and buffer data. Each stream corresponds to a `path` (A path represents a dispatcher), and unprocessed data is stored in its unlimited queue, `pendingQueue`.
```

Comment:
How many stream are there? Why pause the path if exceeds `20%` ?

Reply:
There may be millions of paths in the system. To prevent a single path from consuming all memory resources, we should set a limit: if a path’s memory usage reaches 20%, it should be paused immediately.